### ISSUE TITLE
stdlib becomes detached with Fenix-provided Rust

### ISSUE THREAD

**Issue by charles-dyfis-net** [intellij-rust/intellij-rust#8320]
## Environment

* **IntelliJ Rust plugin version:** 0.4.162.4321-213
* **Rust toolchain version:** 1.59.0-nightly (a359ce160 2021-12-14)
* **IDE name and version:** CLion 2021.3 Build #CL-213.5744.254, built on November 30, 2021
* **Operating system:** macOS Monterey 12.1

## Problem description

I'm using the IntelliJ Rust plugin with CLion on MacOS with Nix as the package manager to install Nix, with the Fenix Rust distro from https://github.com/nix-community/fenix. I have the Rust sources attached to a fixed filesystem location as described by the following code:

```nix
{ pkgs ? import <nixpkgs> {}, name ? "user-env"}:
let
  fenix = import (fetchTarball "https://github.com/nix-community/fenix/archive/main.tar.gz") { };
in with pkgs;
buildEnv {
  inherit name;
  extraOutputsToInstall = ["out" "bin" "lib"];
  paths = [
    fenix.latest.toolchain
    (runCommand "rust-src-link" { rustSrc = fenix.latest.rust-src; } ''
      mkdir -p -- "$out/lib"
      ln -s "$rustSrc" "$out/lib/rust-src"
    '')
  ];
}
```

After saving the above as `~/default.nix` and running `nix-env --set -f ~`, I then have Rust binaries in `~/.nix-profile/bin`, and stdlib sources linked from `~/.nix-profile/lib/rust-src`, which CLion accepts when the following values are in `workspace.xml`:

```xml
    <option name="explicitPathToStdlib" value="$USER_HOME$/.nix-profile/lib/rustlib/src/rust" />
    <option name="toolchainHomeDirectory" value="$USER_HOME$/.nix-profile/bin" />
``` 

(aside: when first setting the bin directory, the stdlib path is autodetected as `/nix/store/akizwak1frfw0xj5zmalr7fc12smadvy-rust-nightly-latest-2021-12-17/lib/rustlib/src/rust`, which is where the `~/.nix-profile/lib/rustlib/src/rust` symlink points)

...at least, the above transiently works (enabling the IDE to figure out types from stdlib code) when my Rust project is first set up in CLion. After I try to build that project, however, the IDE's type detection around values returned by standard-library calls stops working and the "Can not attach stdlib sources automatically without rustup" message that was present before the above settings were added reappears. Clearing caches, restarting the IDE, etc. does not fix this issue once it has happened.

## Steps to reproduce

Given above. Beyond that, there does not appear to be any particular code required. For example, the following:

```rust
use std::io::stdin;
fn main() {
    let stdin = stdin();
    println!("Hello, world!");
}
```

...suffices (when the issue is exhibiting, the IDE says that the type of the variable `stdin` is `<unknown>`).

**Comment by VergeDX** [comment]
Same issues here, this issues seems exists for a long time : (
I tried different Rust overlays on NixOS, it works only on random times...

**Comment by charles-dyfis-net** [comment]
I'd be thrilled just to know where to look for logs to determine what's going wrong when it comes detached (or how to enable such logging should a toggle be needed). I realize the ticket is light on details, but the failure mode is pretty darned opaque; if there were an exception that explained _why_ the previously-attached stdlib sources are being discarded as invalid, that would be hugely helpful.

**Comment by VergeDX** [comment]
I solved this issues by using rustup...
Just adding a rustup update phase in home-manager an keep up-to-date : )

**Comment by charles-dyfis-net** [comment]
_nod_. I hate going the rustup route, though -- it feels like giving up.

More concretely, it means I have (at least) two separate Rust installations (the Nix-provided one and the rustup-maintained one), and that rolling back my Nix profile doesn't actually return to an old system state.

**Comment by VergeDX** [comment]
Right, Act `rustup update` as an activation script of home-manager can be better.

**Comment by charles-dyfis-net** [comment]
I don't see how that's any better. (Also, I use https://gist.github.com/lheckemann/402e61e8e53f136f239ecd8c17ab1deb, not home-manager).
